modules/miner/update.go

```diff
@@ -28,11 +27,138 @@ func (m *Miner) addNewTxns(diff *modules.TransactionPoolDiff) {
+	if m.blockMapHeap.size+candidateSet.size < types.BlockSizeLimit-5e3 {
+		m.blockMapHeap.Push(elem)
+		return
+	}
+
+	// While the heap cannot fit this set s, and while the (weighted) average
+	// fee for the lowest sets from the block is less than the fee for the set
+	// s, continue removing from the heap. The block heap doesn't have enough
+	// space for this transaction. Check if removing sets from the blockMapHeap
+	// will be worth it. bottomSets will hold  the lowest fee sets from the
+	// blockMapHeap
+	bottomSets := make([]*mapElement, 0)
+	var sizeOfBottomSets uint64
+	var averageFeeOfBottomSets types.Currency
+	for {
+		// Check if the candidateSet can fit in the block.
+		if m.blockMapHeap.size-sizeOfBottomSets+candidateSet.size < types.BlockSizeLimit-5e3 {
+			// Place candidate into block,
+			m.blockMapHeap.Push(elem)
+
+			// Places transactions removed from block heap into
+			// the overflow heap.
+			for _, v := range bottomSets {
+				m.overflowMapHeap.Push(v)
+			}
+			break
+		}
+
+		// If the blockMapHeap is empty, push all elements removed from it back
+		// in, and place the candidate set into the overflow. This should never
+		// happen since transaction sets are much smaller than the max block
+		// size.
+		_, exists := m.blockMapHeap.Peek()
+		if !exists {
+			m.overflowMapHeap.Push(elem)
+			// Put back in transactions removed.
+			for _, v := range bottomSets {
+				m.blockMapHeap.Push(v)
+			}
+			// Go to the next candidate set.
+			break
+		}
+		// Add the set to the bottomSets slice. Note that we don't increase
+		// sizeOfBottomSets until after calculating the average.
+		nextSet := m.blockMapHeap.Pop()
+		bottomSets = append(bottomSets, nextSet)
+
+		// Calculating fees to compare total fee from those sets removed and the current set s.
+		totalFeeFromNextSet := nextSet.set.averageFee.Mul64(nextSet.set.size)
+		totalBottomFees := averageFeeOfBottomSets.Mul64(sizeOfBottomSets).Add(totalFeeFromNextSet)
+		sizeOfBottomSets += nextSet.set.size
+		averageFeeOfBottomSets := totalBottomFees.Div64(sizeOfBottomSets)
+
+		// If the average fee of the bottom sets from the block is higher than
+		// the fee from this candidate set, put the candidate into the overflow
+		// MapHeap.
+		if averageFeeOfBottomSets.Cmp(candidateSet.averageFee) == 1 {
+			// CandidateSet goes into the overflow.
+			m.overflowMapHeap.Push(elem)
+			// Put transaction sets from bottom back into the blockMapHeap.
+			for _, v := range bottomSets {
+				m.blockMapHeap.Push(v)
+			}
+			// Go to the next candidate set.
+			break
+		}
+
+	}
+}
+
+// addNewTxns adds new unconfirmed transactions to the miner's transaction
+// selection.
+func (m *Miner) addNewTxns(diff *modules.TransactionPoolDiff) {
+	// Get new splitSets (in form of mapElement)
+	newElements := m.getNewSplitSets(diff)
+
+	// Place each elem in one of the MapHeaps.
+	for i := 0; i < len(newElements); i++ {
+		m.addMapElementTxns(newElements[i])
```

Comment:
this is where I would also add the split sets to the state, you want to do it all together.